### PR TITLE
Set VM USE_VECTOR_REGISTERS flag for supported POWER systems

### PR DESCRIPTION
**Pull request by rmnattas**
Setup the VM USE_VECTOR_REGISTERS flag on supported POWER systems for VectorAPI POWER linkage work.

Signed-off-by: Abdulrahman Alattas <rmnattas@gmail.com>

### CODE REVIEW

**Comment by rmnattas** [comment]
@zl-wang @gita-omr @gacholio

**Review by gacholio — APPROVED** [review]
Note that my assembly changes have yet to be merged.

**Comment by rmnattas** [comment]
@gacholio Should this flag be set only on jdk16 and higher given that the VectorAPI JEB is targeted for that? And if so, would using `#if JAVA_SPEC_VERSION >= 16` be the correct way of doing so?

**Comment by gacholio** [comment]
For now, I'd leave it as-is - my code is not currently ifdeffed.

**Comment by rmnattas** [comment]
Given the build macro required for AIX vector linkage (in 1e4fece) needs the jdk check I added it for the vm flag too. I believe if GAC's changes will use the VM flag it doesn't have to check for the jdk version too.

**Comment by gacholio** [comment]
jenkins compile aix jdknext

**Comment by gacholio** [comment]
@gita-omr Do we want to merge this? Once I've merged my changes, this will incur an immediate penalty on JDK16+ even if no vector code is ever used.

In the future, it may be worth delaying the setting of the bit until the first compilation that actually uses vector registers (this would need to be done under mutex or with an atomic, unlike this code which is guaranteed to be running single-threaded).

**Comment by gacholio** [comment]
At this point, I'm not going to merge this until my changes are in.

**Comment by gacholio** [comment]
My code is in, so I'll run sanity and merge.

**Comment by gacholio** [comment]
jenkins test sanity plinux jdknext

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
@gita-omr If there are performance issues with this when VSRs are not in use, we will need another bit in the flags. The current bit also controls the preservation of the VSRs in the interpreter call-in, which is required in case any method is compiled using VSRs, so it must be set during startup.

**Comment by gita-omr** [comment]
1) this should not be dependent on the JDK level. This is needed even for AutoSIMD which is (at least) in Java 8

**Comment by gita-omr** [comment]
Sorry, I somehow missed the whole discussion on the JDK  being 16 and later. But it definitely was not the intent. We have autosimd bugs where VSRs are not saved by helpers...

**Comment by gita-omr** [comment]
Regarding the vector registers being in use... I think Z somehow knows that but not P. I afraid we will always need to save. Did not we want to do performance runs to see what happens?

**Comment by gacholio** [comment]
I won't merge this as it is.

**Comment by rmnattas** [comment]
I'll update the PR and remove the JDK level check. We didn't performance evaluate @gacholio changes as it wasn't optimized yet, I should be able to evaluate it now before merging this.

**Comment by gacholio** [comment]
FYI my change is causing issues, so I'm going to revert it until I can sort out why.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
Will we need a compiler flag update for Linux as well? I assume today none of the compiled C code uses VSRs at all, so there's no need to preserve them.

**Comment by rmnattas** [comment]
This flag is only required for AIX as it doesn't assume full vector ABI. Both AIX and pLinux already have `-qaltivec` for vector support.

**Comment by zl-wang** [comment]
> OK, this presumably means that we do not support PPC processors without VSR in any OpenJ9 JDK version.

That is certainly true for linux, since Linux LE on POWER started from POWER8.
But this VM flag is still only set after processor testing passes.

**Comment by gacholio** [comment]
Yes, but we're telling the compilers it's OK to use VSRs, so we're assuming processor level (which is fine, just wanted to point it out).

**Comment by gacholio** [comment]
@rmnattas My changes have been merged again, so please proceed with the performance analysis.

I'm going to run some testing on this PR to make sure the vector path is correct (assuming any of the PR build machines meet the processor specs to enable VSR).

**Comment by gacholio** [comment]
jenkins test sanity,extended,sanity.system,extended.system plinux jdk8

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk
Comment (gacholio) [inline runtime/compiler/build/toolcfg/aix-xlc/common.mk:95]: Do we need this flag everywhere (not just the JIT)?

File: runtime/compiler/control/rossa.cpp
Comment (gacholio) [inline runtime/compiler/control/rossa.cpp:1517]: VSR is only supported in 64-bit mode, so this ifdef needs to check that (or some other check needs to be added).

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk
Comment (zl-wang) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk:95]: It is definitely needed where signal handling/longjmp/setjmp/context-related activities can happen.  or, where vr52-63 can potentially be touched (unlikely outside JIT-ed code though).

File: runtime/compiler/control/rossa.cpp
Comment (zl-wang) [reply runtime/compiler/control/rossa.cpp:1517]: hardware-wise and JIT-wise, vector support is available in 32bit too.  maybe from reality perspective, no 32bit build beyond java8. do you mean that?

Comment (gacholio) [reply runtime/compiler/control/rossa.cpp:1517]: It was a somewhat arbitrary decision - the VM helpers only do the VSR work on 64-bit, so we either need to fix that, or the JIT can't use them in 32-bit mode.

Does autoSIMD support 32-bit?

Comment (zl-wang) [reply runtime/compiler/control/rossa.cpp:1517]: > Does autoSIMD support 32-bit?

yes, i believed so.

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk
Comment (gacholio) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk:95]: The actual setjmp/longjmp for exception handling is in OMR. I agree it's extremely unlikely that any generic C code in the VM will need to use VSR at all, let alone the top-end ones, but if it's possible, then we'll need the define. The JCL code will also need to have it.

File: runtime/compiler/control/rossa.cpp
Comment (gacholio) [reply runtime/compiler/control/rossa.cpp:1517]: @gita-omr Do we want support on 32-bit (bearing mind the continued existence of the 32-bit IBM builds)?

**Comment by gacholio** [comment]
All testing passed, so either the new code is funcitonally correct or none of the machines support VSR (which seems unlikely at this point). I'm doing some internal testing as well to verify.

File: runtime/compiler/control/rossa.cpp
Comment (gita-omr) [reply runtime/compiler/control/rossa.cpp:1517]: Yes, we never treated 32-bit any differently in autoSIMD. I don't think we want to disable it now. Sorry, I probably should've asked why we only save vector registers in 64-bit mode...

**Comment by gacholio** [comment]
@pshipton You recently mentioned continued support for Power 6 - I believe the compiler options currently in use (`-qaltivec`) and possibly the one being proposed here allow the compiler to use instructions that are not supported on Power 6.

**Comment by zl-wang** [comment]
need to take a deeper look at P6.  altivec (aka VMX) is supported by P6, but VSX is not.  -qaltivec causing a crash of (SIGILL?) sounds to me something unexpected. maybe the AIX compiler took it to mean VSX too.

**Comment by gacholio** [comment]
The crashes we were experiencing were due to me testing the wrong field for the vector enable flag, so we were taking the VSX path incorrectly. This is fixed by the new commit in #12350.

File: runtime/compiler/control/rossa.cpp
Comment (gacholio) [reply runtime/compiler/control/rossa.cpp:1517]: I will address this in a new PR shortly.

**Comment by gacholio** [comment]
Would it make sense not to use the EXTABI flag? There's no reason to save the VSRs if they aren't being used by the C code, and there's no good reason any of our code would need to use them. Not introducing the flag maintains the Power6 backwards compatibility, I believe.

File: runtime/compiler/control/rossa.cpp
Comment (gacholio) [reply runtime/compiler/control/rossa.cpp:1517]: We've opted to defer 32-bit support, but it doesn't matter to the VM if the bit is set in 32-bit mode (all of the VM code is ifdeffed on 64-bit). The autosimd code will need to be aware that helpers only preserve VSRs on 64-bit.

**Comment by gacholio** [comment]
Based on the conversation above, it appears the compiler flag may be needed for signal handling, but we'll need more detail to decide where/if this flag should be defined. The setjmp/longjmp code will presumably handle the different processor capabilities.

**Comment by gacholio** [comment]
@rmnattas Any word on the performance? I believe the VM-side code is finally in to stay.

**Comment by zl-wang** [comment]
without EXTABI, it is unclear the implication of signal handling (in JIT-ed code in particular). i supposed it doesn't work (if vector registers are actively used at the point).

the preserved set of VSR/VR existed on P6.  xlC would not generate VSX (but VMX) instructions to handle them I would assume.

**Comment by gacholio** [comment]
I don't see any changes in the last push.

**Comment by rmnattas** [comment]
Wasn't sure if it notifies for this. Yes, it's just a rebase. Working on evaluating the performance.

**Comment by rmnattas** [comment]
Comparing performance of a build without the changes in #12350, and another with #12350 + #12046 (this PR to enable the changes), the changes results in %34 more time spent in saving/restoring the registers.

In my basic microbench that emphasize calling helpers that will use the save/restore registers that got changed many times, it's still not a large portion of the run. The save/restore registers portion increased from being 0.91% (~4283 perf samples) of the run to 1.22% (~5777 perf samples). Resulting in 0.8% increase in runtime.

**Comment by zl-wang** [comment]
that number makes sense more or less.  normal application probably uses these code(s) sporadically only, such that it is not a concern.

**Comment by gacholio** [comment]
Given that helpers should not be called in any performance-critical path (with the one exception of interface dispatch, which does not need to save the VRs), I think this means we can enable the feature on 64-bit (which needs a change here).

There is also the matter of the compiler options and where they need to be applied.

**Comment by rmnattas** [comment]
Added 64bit checks for the flags. @gacholio @zl-wang @gita-omr

**Comment by gacholio** [comment]
I believe the  compiler option is unnecessary for the JIT code (if anything, the OMR signal handling code needs it), so it should be removed from this PR and done separately (if at all).

**Comment by rmnattas** [comment]
Removed the compiler option here. Will see for OMR and where to add the option.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins test sanity plinux,aix jdk8

**Comment by gacholio** [comment]
@gita-omr Do you approve merging this?

**Comment by gita-omr** [comment]
Jenkins test sanity aix,plinux jdk8,jdk11

**Comment by gita-omr** [comment]
Sorry, did not notice you already ran the sanity...

**Comment by gita-omr** [comment]
@zl-wang do you think we can accept the performance impact, if any?

**Comment by zl-wang** [comment]
@gita-omr  i answered here already:  https://github.com/eclipse/openj9/pull/12046#issuecomment-824245752

**Review by gita-omr — APPROVED** [review]

**Pull request merged**
